ext/mbstring/libmbfl/mbfl/mbfl_encoding.h

```diff
@@ -129,10 +130,110 @@ struct mbfl_convert_vtbl {
 	void (*filter_copy)(struct _mbfl_convert_filter *src, struct _mbfl_convert_filter *dest);
 };
 
-/*
- * encoding
- */
-typedef struct _mbfl_encoding {
+typedef struct {
+	uint32_t state;
+	uint32_t errors;
+	zend_ulong maxsize;
+	size_t len;
+	char   val[1];
+} mb_convert_buf;
+
+typedef size_t (*mb_to_wchar_fn)(unsigned char **in, size_t *in_len, uint32_t *out, size_t out_len, unsigned int *state);
+typedef mb_convert_buf* (*mb_from_wchar_fn)(uint32_t *in, size_t len, mb_convert_buf *out, bool end);
+
+static inline void mb_convert_buf_init(mb_convert_buf *buf, size_t initsize)
+{
+	buf->state = buf->errors = buf->len = 0;
+	buf->maxsize = initsize;
+}
+
+static inline mb_convert_buf* mb_convert_buf_alloc(size_t initsize)
+{
+	mb_convert_buf *buf = emalloc(sizeof(mb_convert_buf) + initsize);
+	mb_convert_buf_init(buf, initsize);
+	return buf;
+}
+
+static inline mb_convert_buf *mb_convert_buf_ensure(mb_convert_buf *buf, size_t needed)
+{
+	if (buf->maxsize < needed) {
+		size_t newlen = MAX(buf->maxsize + (buf->maxsize >> 1), needed);
+		buf = erealloc(buf, sizeof(mb_convert_buf) + newlen);
+		buf->maxsize = newlen;
+	}
+	return buf;
+}
+
+static inline void mb_convert_buf_add(mb_convert_buf *buf, char c)
+{
+	buf->val[buf->len++] = c;
```

Comment:
Assert that it fits in max size in these?

Reply:
I did that while testing the code. Just benchmarked and confirmed it doesn't add any measurable overhead... so into production it goes.